segment/writer.go

```diff
@@ -257,6 +257,26 @@ func (w *Writer) Append(entries []types.LogEntry) error {
 		return types.ErrSealed
 	}
 
+	flushed := false
+
+	// Save any state we may need to rollback.
+	beforeBuf := w.writer.commitBuf
+	beforeCRC := w.writer.crc
+	beforeIndexStart := w.writer.indexStart
+	beforeWriteOffset := w.writer.writeOffset
+	beforeOffsets := w.offsets.Load()
+
+	defer func() {
+		if !flushed {
```

Comment:
I considered it but I felt like it there were too many footguns because it relies on the code and all future editors being very careful about not shadowing `err`. For example the code below:

```go
if err := w.appendEntry(e); err != nil {
  return err
}
```

This wouldn't set the outer scope err correctly and so would be a really easy way to accidentally mis a case.

It felt more readable and less footgunny to name it explicitly!